fedbiomed/node/requests/_pending_requests.py

```diff
@@ -0,0 +1,196 @@
+                'event': threading.Event(),
+                'requests': request_ids,
+            }
+
+        return self._id_counter
+
+
+    def wait(self, listener_id: int, timeout: float) -> Tuple[bool, List[InnerMessage]]:
+        """Wait for a registered listener to complete.
+
+        Blocks until all replies for this listener are received or until timeout is reached.
+
+        Args:
+            listener_id: unique ID of this listener
+            timeout: maximum time to wait for replies in seconds
+
+        Returns:
+            A tuple consisting of
+                - a bool set to `True` if all replies for this listener were received, `False` if not
+                - a list of the messages for the received replies
+
+        Raises:
+               FedbiomedNodeToNodeError: timeout has incorrect type or value
+        """
+
+        # Check value for timeout, as bad value may cause hard to detect problems
+        if (not isinstance(timeout, float) and not (isinstance(timeout, int))) \
+                or timeout < 0 or timeout > MAX_PENDING_REPLY_TIMEOUT:
+            raise FedbiomedNodeToNodeError(f"{ErrorNumbers.FB324}: Cannot wait {timeout} seconds. "
+                                           f"Should be float between 0 and {MAX_PENDING_REPLY_TIMEOUT}")
+
+        time_initial = time.time()
+
+        with self._pending_listeners_lock:
+            # Check a listener exists for this id
+            if listener_id not in self._pending_listeners:
+                # don't raise exception, non fatal error
+                logger.error(f"{ErrorNumbers.FB324}: Node tries to wait for replies "
+                             f"of a non existing request-reply: {listener_id}")
+                return False, []
+
+            # Retrieve event object for this listener
+            event = self._pending_listeners[listener_id]['event']
+
+        # wait until all replies are received or timeout is reached
+        while not self._all_replies(listener_id) and (time.time() < time_initial + timeout):
+            # be sure not to hold any lock when waiting !
+            event.wait(time_initial + timeout - time.time())
+            event.clear()
+
+        with self._pending_listeners_lock:
+            with self._pending_replies_lock:
+                # check if all replies were received
+                all_received = set(self._pending_listeners[listener_id]['requests']).\
+                    issubset(set(self._pending_replies.keys()))
+
+                # remove all replies for this request from the pending replies,
+                # and add them to the list of received replies
+                messages = [self._pending_replies.pop(reqid)
+                            for reqid in self._pending_listeners[listener_id]['requests']
+                            if reqid in self._pending_replies]
+
+                # remove expired listener
+                del self._pending_listeners[listener_id]
```

Comment:
Same comment for this part as well, however, I don't know if you address the issue when "wait" is called twice using same `listener_id`. If it is the case yes using locks  will make it safe. 

https://docs.python.org/3/faq/library.html#what-kinds-of-global-value-mutation-are-thread-safe

Reply:
Thank you @srcansiz 

As above, we have multiple operations in sequence, so even if one is atomic the combination is not.
Moreover, we might have `_clean_pending_replies` removing entries in the while.

So I think this is really needed.